### PR TITLE
Wake agents when async bash commands finish

### PR DESCRIPTION
**Pull request by sethkarten**
## Context

No-Ticket: Requested follow-up behavior for the async kernel `bash()` API.

An agent can start a long shell command and end its turn, but nothing currently wakes it when the command finishes. This leaves the agent dependent on manual polling even though the command already has a managed handle and lifecycle.

## Changes

- Send a `bash.completed` host notification when a handle left running beyond its creating cell finishes.
- Queue a model-visible follow-up with the PID, exit code, command, and handle inspection methods.
- Suppress the follow-up when the creating cell directly or transitively awaits the handle, while detached task awaits still wake the agent.
- Attribute completion barriers to unique per-cell execution contexts so reused protocol request IDs cannot delay old notifications.
- Render async shell completion messages in the transcript and document the behavior.

## Validation

- `npm run check`
- `env -u RLM_MAX_DEPTH npx tsx ../../node_modules/vitest/dist/cli.js --run test/async-bash-completion.test.ts test/agent-session-recursion.test.ts test/system-prompt.test.ts` (146 passed)
- `uv run python -m unittest discover -s test` (273 passed)
- `uv run python -m unittest test.test_bash` (56 passed)

---

> [!NOTE]
> **Medium Risk**
> Wake/suppress logic depends on heuristic asyncio await-chain detection in the runtime; edge-case patterns could miss or duplicate notifications, affecting agent orchestration but not security or data handling.
> 
> **Overview**
> Agents can start long `bash()` jobs, end their turn, and get **automatic follow-ups** when those shells exit instead of polling blindly.
> 
> The Python runtime tracks each REPL cell’s completion barrier and, after a **detached** `BashHandle` finishes, sends a `bash.completed` host request (PID, command, exit code). Notifications are **suppressed** when the creating cell directly or transitively awaits the handle (including common `asyncio` wrappers); awaits from separate background tasks still trigger wake-ups. Per-cell execution context avoids stale notifications when protocol request IDs are reused.
> 
> The coding agent validates `bash.completed`, injects an **`async_bash_completion`** custom message, and queues/resumes a model turn with instructions to inspect the saved handle via `.poll()`, `.output()`, or `.tail()`. Docs, RLM prompts, and the interactive transcript reflect the new behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1bb4d43a0a64366217f8fec38d574e3917fa550d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Wake agents when background bash commands finish
> - When a REPL cell starts a long-running bash command and finishes without awaiting it, the `BashHandle` sends a `bash.completed` host request with the PID, command, and exit code.
> - The agent session handles `bash.completed` by creating an injected custom message and queuing a follow-up to resume the session.
> - `rlm.createAsyncBashCompletionHostHandler` validates completion requests, rejecting invalid PIDs, empty commands, and non-integer exit codes before invoking the completion callback.
> - Risk: `BashHandle._notify_background_completion` in [bash.py](https://github.com/PrimeIntellect-ai/prime-agent/pull/2053/files#diff-f2d79e9ee0bfb6f41372db8ae482096e4f76ba1d03444cfe4b90cd0b0b599287) ignores `OSError` and `RuntimeError` from the bridge. If an asyncio wrapper path is not recognized by `_completion_reaches`, the handle may be incorrectly treated as awaited and not notify the agent.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 1bb4d43.</sup>
> 
>

### CODE REVIEW

File: prime-agent-runtime/src/rlm/bash.py [hunk 1]
```diff
@@ -515,6 +537,45 @@ def _add_done_callback(self, callback: Callable[[], None]) -> None:
...
+            return
+        task = loop.create_task(self._notify_background_completion(cell_finished))
+        task.add_done_callback(_consume_notice_task)
+
+    async def _notify_background_completion(self, cell_finished: asyncio.Event) -> None:
+        result = await self._wait()
+        # The cell may do other work before awaiting this handle. Do not classify
+        # it as detached until that whole cell has crossed its completion barrier.
+        await cell_finished.wait()
+        if self._await_started:
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply prime-agent-runtime/src/rlm/bash.py]: Fixed in 305f88fa9. BashHandle now records the REPL cell owner task and suppresses the completion notice only when that task awaits the handle. A regression test covers a detached asyncio task awaiting the same handle while the creating cell exits.

File: prime-agent-runtime/src/rlm/repl.py [hunk 2]
```diff
@@ -98,6 +102,28 @@ def is_active() -> bool:
...
+
+    Detached asyncio tasks inherit the creating cell id. If such a task starts
+    after that cell finished, return an already-open barrier instead of retaining
+    completed request ids indefinitely.
+    """
+    cell_id = _current_cell.get()
+    if cell_id is None or _loop is None:
+        return None
+    with _interrupt_lock:
+        if cell_id not in _inflight:
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply prime-agent-runtime/src/rlm/repl.py]: Fixed in 305f88fa9. Completion barriers are now keyed by a monotonic internal cell generation carried in context, independent of caller-supplied request IDs. A regression test reuses an ID and verifies the old-cell completion is not captured by the newer cell.

File: prime-agent-runtime/src/rlm/bash.py [hunk 3]
```diff
@@ -515,6 +537,45 @@ def _add_done_callback(self, callback: Callable[[], None]) -> None:
...
+            return
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply prime-agent-runtime/src/rlm/bash.py:643]: Fixed in 305f88fa9. Await suppression is now scoped to the creating cell owner task, so later or cancelled detached-task awaits do not disable the wake-up. The new detached-task await regression test covers this path.

File: prime-agent-runtime/src/rlm/bash.py [hunk 4]
```diff
@@ -639,6 +702,12 @@ def __await__(self) -> Generator[Any, None, BashResult]:
         # A handle awaited before any other API use is a one-shot command tied
         # to the await (kill-on-cancel); touching the handle API first marks it
         # as a deliberate background handle whose awaits only wait.
+        try:
+            current_task = asyncio.current_task()
+        except RuntimeError:
+            current_task = None
+        if current_task is not None and current_task is self._creating_cell_task:
+            self._awaited_by_creating_cell = True
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply prime-agent-runtime/src/rlm/bash.py:794]: Fixed in ffb28fc95. The runtime now follows the creating cell task's active asyncio dependency chain, including gather, wait_for, shield, and nested wrappers, before deciding that an await was detached. Regression coverage runs these wrapper forms on Python 3.13, with focused Python 3.11 validation as well; a detached task awaiting the handle while the cell waits on unrelated work still emits the completion.

File: prime-agent-runtime/src/rlm/bash.py [hunk 5]
```diff
@@ -43,13 +44,94 @@
...
+    return None
+
+
+def _consume_notice_task(task: asyncio.Task[None]) -> None:
+    """Retrieve detached notifier failures so they never become loop warnings."""
+    if not task.cancelled():
+        task.exception()
+
+
+def _completion_reaches(start: asyncio.Future[Any], target: asyncio.Future[Any]) -> bool:
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply prime-agent-runtime/src/rlm/bash.py]: Fixed in 1eab8bc51. `_completion_reaches` now ignores arbitrary done callbacks and follows only explicit `gather`, `shield`, `wait_for`, and `TaskGroup` ownership edges. Added regressions for a creating-cell `TaskGroup` await and for an unrelated callback closure capturing the cell waiter; both pass on Python 3.11 and 3.13.

File: prime-agent-runtime/src/rlm/bash.py [hunk 6]
```diff
@@ -43,13 +44,110 @@
 # wait for a confirmed group exit before CancelledError propagates.
 _CANCEL_TERM_GRACE = 0.5
 _CANCEL_KILL_WAIT = 2.0
+_COMPLETION_NOTICE_COMMAND_CAP = 1000
+_ASYNCIO_WRAPPER_CALLBACKS = {
+    ("asyncio.tasks", "gather.<locals>._done_callback"),
+    ("asyncio.tasks", "shield.<locals>._inner_done_callback"),
+    ("asyncio.tasks", "_release_waiter"),
+}
```
[the remark this answers is not in the document]
Comment (sethkarten) [reply prime-agent-runtime/src/rlm/bash.py:53]: Fixed in 1bb4d43a0. The ownership graph now includes `asyncio.wait` completion callbacks and the concrete `Task.task_wakeup` edge, which connects nested `gather`/`shield` futures back through a `TaskGroup` child to its parent cell. Added Python 3.11 and 3.13 regressions for `wait`, direct TaskGroup waits, and TaskGroup children using both wrappers.

**Comment by sethkarten** [comment]
Included in #2068: https://github.com/PrimeIntellect-ai/prime-agent/pull/2068
